README.md
# Clint's Python Utilities

- ![Tests](https://github.com/clintecker/python-clint-utilities/workflows/Tests/badge.svg)
- ![Coverage](https://github.com/clintecker/python-clint-utilities/workflows/Coverage/badge.svg)


CHANGELOG
# 1.2.2 

Stopped disabling other loggers.

# 1.2.0

Added caching to test runs, badges on the README, documented all the functions, ensured
all the functions hade proper types. Changed many of the date-related functions which
breaks the calling contract.

# 1.1.1

Incorrectly versioned things again so this version should get everything back in sync.

# 1.1.0

Changed the calling signautre of the `setupLogging` function and started testing on Python 3.9

# 1.0.4

Removal of dead code.

# 1.0.3

More logging fixes.

# 1.0.2

Making logging setup functional

# 1.0.1

Adding a logger setup function

# 1.0.0

Initial release!


tests/test_utilities.py
# Standard Library
import datetime
import json
from unittest import mock

# Third Party Code
from dateutil.tz import tzoffset
import requests
import responses

# Clint Utilities Code
from clint_utilities import (
    _PARSED_DATES,
    assert_raises,
    date_string_to_date_timestamp,
    datetime_to_date_timestamp,
    make_durable_get,
    make_durable_post,
    make_durable_request,
    parse_args,
    parse_date_string,
    parse_utc_timestamp,
    RequestError,
    RequestSuccess,
    setup_logging,
)


def test_parse_date():
    assert len(_PARSED_DATES) == 0
    assert datetime.datetime(
        2020, 6, 20, 14, 39, 10, tzinfo=tzoffset(None, -21600)
    ) == parse_date_string("2020-06-20T14:39:10-0600")
    assert len(_PARSED_DATES) == 1
    # Test caching
    assert datetime.datetime(
        2020, 6, 20, 14, 39, 10, tzinfo=tzoffset(None, -21600)
    ) == parse_date_string("2020-06-20T14:39:10-0600")
    assert len(_PARSED_DATES) == 1


def test_parse_utc_timestamp():
    assert datetime.datetime(
        2020, 7, 7, 15, 1, 56, tzinfo=tzoffset("Local", -21600)
    ) == parse_utc_timestamp(ts=1594155716, offset=-21600)
    assert 1594155716 == parse_utc_timestamp(ts=1594155716, offset=-21600).timestamp()


def test_datetime_to_timestamp():
    assert 1592611200 == datetime_to_date_timestamp(
        datetime.datetime(2020, 6, 20, 14, 39, 10)
    )


def test_date_string_to_timestamp():
    assert 1592611200 == date_string_to_date_timestamp("2020-06-20")


def test_make_durable_request_num_attempts_value_error():
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.get, url="http://example.com", num_attempts=0, delay=1.0
        ),
        args=(),
        exc=ValueError,
    )


def test_make_durable_request_delay_value_error():
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.get, url="http://example.com", num_attempts=3, delay=-1.0
        ),
        args=(),
        exc=ValueError,
    )


@responses.activate
def test_make_durable_request_without_json():
    responses.add(
        method=responses.GET,
        url="http://google.com/",
        status=200,
        body="Cool.",
        headers={"Content-Type": "text/plain"},
    )
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.get,
            url="http://google.com/",
            success_codes=[200,],
            num_attempts=2,
            delay=2.0,
        ),
        args=(),
        exc=RequestSuccess,
    )


@responses.activate
def test_make_durable_request_1_bad_response():
    responses.add(method=responses.POST, url="https://example.com/resource", status=504)
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.post,
            url="https://example.com/resource",
            num_attempts=1,
            delay=5,
            json={"resources": [{"id": 1}, {"id": 2}, {"id": 3}]},
        ),
        args=(),
        exc=RequestError,
    )


@responses.activate
def test_make_durable_request_succeeds_after_retry():
    responses.add(method=responses.POST, url="https://example.com/resource", status=504)
    responses.add(method=responses.POST, url="https://example.com/resource", status=201)
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.post,
            url="https://example.com/resource",
            num_attempts=3,
            delay=0,
            json={"resources": [{"id": 1}, {"id": 2}, {"id": 3}]},
        ),
        args=(),
        exc=RequestSuccess,
    )


@responses.activate
def test_make_durable_request_client_error():
    responses.add(method=responses.POST, url="https://example.com/resource", status=400)
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.post,
            url="https://example.com/resource",
            num_attempts=3,
            delay=0,
            json={"resources": [{"id": 1}, {"id": 2}, {"id": 3}]},
        ),
        args=(),
        exc=RequestError,
    )


@responses.activate
def test_make_durable_request_connection_error():
    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=requests.post,
            url="https://example.com/resource",
            num_attempts=3,
            delay=0,
            json={"resources": [{"id": 1}, {"id": 2}, {"id": 3}]},
        ),
        args=(),
        exc=RequestError,
    )


def test_make_durable_request_unhandled_error():
    def method(*args, **kwargs):
        raise ValueError()

    assert_raises(
        fn=make_durable_request,
        kwargs=dict(
            method=method,
            url="http://127.0.0.1:8000/resource",
            num_attempts=3,
            delay=0,
            json={"resources": [{"id": 1}, {"id": 2}, {"id": 3}]},
        ),
        args=(),
        exc=ValueError,
    )


@responses.activate
def test_make_durable_post():
    responses.add(
        url="http://example.com/resource",
        method=responses.POST,
        status=201,
        body=json.dumps({"data": {"resource": {"id": 1}}}),
        headers={"Content-Type": "application/json"},
    )
    response = make_durable_post(
        "http://example.com/resource", 3, 1.0, json={"resource": {"id": 1}}
    )
    assert response.json() == {"data": {"resource": {"id": 1}}}


@responses.activate
def test_make_durable_get():
    responses.add(
        url="http://example.com/resource/1",
        method=responses.GET,
        status=200,
        body="Cool",
        headers={"Content-Type": "text/plain"},
    )
    response = make_durable_get("http://example.com/resource/1", 3, 1.0)
    assert "Cool" == response.text


@mock.patch("clint_utilities.argparse.ArgumentParser")
def test_parse_args(mock_arg_parse_cls):
    mock_parser = mock.Mock(name="parser")
    mock_arg_parse_cls.return_value = mock_parser
    parse_args(
        {
            "description": "This is what my tool does!",
            "args": {
                "--option,-O": {
                    "dest": "option_cool",
                    "type": float,
                    "default": 0.00,
                    "help": "Please input the option value!",
                },
                "--flag,-F": {
                    "dest": "mean_flag",
                    "type": bool,
                    "action": "store_true",
                    "default": False,
                    "help": "Do you want to be mean??",
                },
            },
        }
    )
    assert mock_arg_parse_cls.called_with(description="This is what my tool does!")
    assert [
        mock.call.add_argument(
            "--option",
            "-O",
            dest="option_cool",
            type=float,
            default=0.0,
            help="Please input the option value!",
        ),
        mock.call.add_argument(
            "--flag",
            "-F",
            dest="mean_flag",
            type=bool,
            action="store_true",
            default=False,
            help="Do you want to be mean??",
        ),
        mock.call.parse_args(),
    ] == mock_parser.method_calls


def test_assert_raises():
    def oops():
        raise ValueError

    def oops2():
        raise TypeError

    def oops3():
        return 3

    assert_raises(fn=oops, args=tuple(), kwargs=dict(), exc=ValueError)

    assert_raises(
        fn=assert_raises,
        args=tuple(),
        kwargs=dict(fn=oops2, args=tuple(), kwargs={}, exc=ValueError),
        exc=AssertionError,
    )

    assert_raises(
        fn=assert_raises,
        args=tuple(),
        kwargs=dict(fn=oops3, args=tuple(), kwargs={}, exc=ValueError),
        exc=AssertionError,
    )


@mock.patch("clint_utilities.dictConfig")
def test_setup_logging(mock_dict_config):
    setup_logging(
        "DEBUG",
        msg_format="%(asctime)s: %(message)s",
        formatter_class_name="project.utils.RequestFormatter",
    )
    mock_dict_config.assert_called_once_with(
        {
            "version": 1,
            "formatters": {
                "default": {
                    "format": "%(asctime)s: %(message)s",
                    "class": "project.utils.RequestFormatter",
                }
            },
            "handlers": {
                "console": {
                    "level": "DEBUG",
                    "class": "logging.StreamHandler",
                    "formatter": "default",
                    "stream": "ext://sys.stdout",
                },
                "file": {
                    "level": "DEBUG",
                    "class": "logging.handlers.RotatingFileHandler",
                    "formatter": "default",
                    "filename": "error.log",
                    "maxBytes": 1073741824,
                },
            },
            "root": {
                "level": "DEBUG",
                "handlers": ["console", "file"],
                "propagate": True,
            },
            "disable_existing_loggers": True,
        }
    )


@mock.patch("clint_utilities.dictConfig")
def test_setup_logging_no_request_class(mock_dict_config):
    setup_logging(
        "DEBUG", msg_format="%(asctime)s: %(message)s",
    )
    mock_dict_config.assert_called_once_with(
        {
            "version": 1,
            "formatters": {"default": {"format": "%(asctime)s: %(message)s"}},
            "handlers": {
                "console": {
                    "level": "DEBUG",
                    "class": "logging.StreamHandler",
                    "formatter": "default",
                    "stream": "ext://sys.stdout",
                },
                "file": {
                    "level": "DEBUG",
                    "class": "logging.handlers.RotatingFileHandler",
                    "formatter": "default",
                    "filename": "error.log",
                    "maxBytes": 1073741824,
                },
            },
            "root": {
                "level": "DEBUG",
                "handlers": ["console", "file"],
                "propagate": True,
            },
            "disable_existing_loggers": True,
        }
    )


clint_utilities/__init__.py
# TODO: Almost all the functions in this file need
# documentation, typing information, and cleaned-up
# names

# Standard Library
import argparse
import datetime
import logging
from logging.config import dictConfig
import socket
import time
from typing import Any, Callable, Dict, Iterable, List, Optional, Type

# Third Party Code
from dateutil.parser import parse
from dateutil.tz import tzoffset, tzutc
import requests


VERSION = "1.2.2"
__version__ = VERSION
version = [1, 2, 2]

_PARSED_DATES: Dict[str, datetime.datetime] = {}


logger = logging.getLogger(__name__)


class AlreadyInserted(Exception):
    """
    Indicates that the data was previously inserted
    """

    pass


class RequestError(Exception):
    """
    Indicates that the request failed.
    """

    pass


class RequestSuccess(Exception):
    """
    Indicates that the request succeeded.

    :param response: The response to the successful request.
    """

    def __init__(self, response: requests.Response):
        self.response = response
        super(RequestSuccess, self).__init__()


class StoreSuccess(RequestSuccess):
    """
    Indicates that the storage request was successfully completed.
    """

    pass


class StoreError(RequestError):
    """
    Indicates that the storage request failed.
    """

    pass


def parse_utc_timestamp(ts: int, offset: int) -> datetime.datetime:
    """
    Converts an epoch-base timestamp and offset to a timezone-aware datetime object.

    :param ts: An integer representing the number of seconds since Jan 1, 1970 UTC
    :param offset: An integer indicating the number of seconds west (negative) or west (positive) of the date line.
    :returns: A timezone aware datetime object representing the moment specified in the timestamp.
    """
    return (
        datetime.datetime.utcfromtimestamp(ts)
        .replace(tzinfo=datetime.timezone.utc)
        .astimezone(tzoffset(name="Local", offset=offset))
    )


def parse_date_string(date_string: str) -> datetime.datetime:
    """
    Parses a string containing a date and produces a (potentially time-zone aware) datetime object.

    :returns: A (potentially time-zone aware) datetime object representing the moment specified in the string.
    :param date_string: A string representing a date and time.
    """
    if _PARSED_DATES.get(date_string) is None:
        _PARSED_DATES[date_string] = parse(date_string)
    return _PARSED_DATES[date_string]


def datetime_to_date_timestamp(dt: datetime.datetime) -> int:
    """
    Converts a datetime object to a timestamp representing the first second of the day specified in the input.

    Given a datetime object representing any moment in time, return an integer timestamp representing the first
    second of the day of the moment.

    :param dt: Any datetime object. If the object is not tz-aware, we will presume UTC.
    :returns: An integer timestamp representing the first second of the day of the input datetime, in UTC.
    """
    tz = dt.tzinfo or tzutc()
    return int(
        datetime.datetime(
            year=dt.year,
            month=dt.month,
            day=dt.day,
            hour=0,
            minute=0,
            second=0,
            tzinfo=tz,
        ).timestamp()
    )


def date_string_to_date_timestamp(date_string: str) -> int:
    """
    Convert a date string to a date timestamp.

    Input could be any kind of date or datetime string, but only the date portion will be extracted
    and used.

    :param date_string: A string representing a date. If a datetime is specified, only the date portion will be used.
    """
    parsed_date = parse_date_string(date_string)
    return datetime_to_date_timestamp(parsed_date)


def make_durable_request(
    method: Callable[..., requests.Response],
    url: str,
    num_attempts: int,
    delay: float,
    success_codes: Optional[List[int]] = None,
    success: Optional[Type[RequestSuccess]] = None,
    error: Optional[Type[RequestError]] = None,
    json: Optional[Dict] = None,
) -> None:
    """
    Make a durable HTTP request.

    Durable means that efforts will be made to ensure that the requested operation has been completed
    despite any kind of transient failures.

    The caller of this method should be able to know definitively whether or not their requested operation
    actually succeeded or will never succeed.

    If the request is made, but network or socket errors occur, a delay will in incurred, and the operation
    will be attempted again until the maximum number of tries is exceeded or the operation succeeds.

    If a server error (5XX) occurs, the retry mechanism will similarly engage.

    Any other kind of error will result in an abortion of the retry mechanism, since subsequent
    attempts at completing the request would be futile.

    :param method: A function that returns a `requests.Response` object.
    :param url: An HTTP URL to make a request to.
    :param num_attempts: The maximum number of requests to attempt before giving up.
    :param delay: The amount of time to delay (in seconds), before making subsequent requests.
    :success_codes: An optional list of HTTP codes that should be considered "successful"
                    (default: [200, 201, 202, 204]).
    :success: An optional exception to raise in the case of success. (default: `RequestSuccess`).
    :error: An optional exceptio nto raise in the case of failure (default: `RequestError`).
    :json: An optional payload to convert to JSON and include in the body when making the request.
    :returns: None
    :raises: Different exceptions depending on whether or not the operation succeeded or failed. An exception
             signalling success will carry a `response` property bearing the entire response object.

    """
    if num_attempts <= 0:
        raise ValueError("The value of `num_attempts` must be greater than 0.")
    if delay < 0:
        raise ValueError("The value of `delay` must be greater than or equal to 0.")
    success = success or RequestSuccess
    error = error or RequestError
    success_codes = success_codes or [200, 201, 202, 204]
    response: Optional[requests.Response] = None
    # Execute request.
    try:
        if json:
            response = method(url=url, json=json)
        else:
            response = method(url=url)
    except (requests.exceptions.ConnectionError, socket.timeout):
        logger.warning("Network error.")

    # Process outcomes.
    if response is not None:
        if response.status_code >= 500 and response.status_code < 600:
            logger.warning("Server error.")
        elif response.status_code in success_codes:
            logger.info("Request was successful.")
            raise success(response=response)
        else:
            logger.error(
                "Could not complete request due to status code %s: %s",
                response.status_code,
                response.text,
            )
            raise error(
                "Could not complete request due to stauts code %s: %s"
                % (response.status_code, response.text)
            )

    if num_attempts > 1:
        logger.info("Will retry request.")
        time.sleep(delay)
        make_durable_request(
            method=method,
            url=url,
            num_attempts=(num_attempts - 1),
            delay=delay * 2,
            success_codes=success_codes,
            success=success,
            error=error,
            json=json,
        )
    else:
        raise error("Unable to complete request.")


def make_durable_post(
    url: str, num_attempts: int, delay: float, json: Dict
) -> requests.Response:
    """
    Make a durable POST operation.

    First read the documentation to `make_durable_request` as this method leverages that code.

    We provide all the user-supplied data, but prefill `method`, `success_codes`, `success,` and `error`
    to provide an nice developer experience for HTTP POST requests.

    :param url: The URL to make the request to.
    :param num_attempts: The maximum number of attempts to make in the face of transient errors.
    :param delay: How many seconds to delay between retries.
    :param json: A payload object that will be converted to JSON and used as the body of the POST request.
    :returns: An HTTP response object.
    """
    try:
        make_durable_request(
            method=requests.post,
            url=url,
            num_attempts=num_attempts,
            delay=delay,
            success_codes=[201, 409],
            success=StoreSuccess,
            error=StoreError,
            json=json,
        )
    except StoreSuccess as exc:
        response = exc.response
    return response


def make_durable_get(url: str, num_attempts: int, delay: float) -> requests.Response:
    """
    Make a durable GET operation.

    First read the documentation to `make_durable_request` as this method leverages that code.

    We provide all the user-supplied data, but prefill `method` to provide an nice developer
    experience for HTTP GET requests.

    :param url: The URL to make the request to.
    :param num_attempts: The maximum number of attempts to make in the face of transient errors.
    :param delay: How many seconds to delay between retries.
    :returns: An HTTP response object.
    """
    try:
        make_durable_request(
            method=requests.get, url=url, num_attempts=num_attempts, delay=delay,
        )
    except RequestSuccess as exc:
        response = exc.response
    return response


def parse_args(configuration: Dict) -> argparse.Namespace:
    """
    Parses program arguments according to a configuration dictionary

    The configuration looks a little bit like the following. Keys for the
    "args" dictionary can contain either only a short or long argument
    name, or can supply both but placing a comma in between them.

    {
        "description": "My program",
        "args": {
            "-C,--color": {
                "type": "int",
                "help": "The color to use"
                "default": 255
            },
            "--name": {
                "type": "str",
                "help": "Your name",
            }
        }
    }

    :param configuration: A dictionary that matches the format above.
    :returns: A completed parsed argument namespace.
    """
    description: str = configuration["description"]
    args: Dict[str, Dict] = configuration["args"]
    parser = argparse.ArgumentParser(description=description)
    for arg, arg_data in iter(args.items()):
        parser.add_argument(*arg.split(","), **arg_data)
    return parser.parse_args()


def assert_raises(
    fn: Callable, args: Iterable[Any], kwargs: Dict[str, Any], exc: Type[Exception]
) -> None:
    """
    Asserts that a specific exception was raised while executing a function.

    If the exception is raised, this method returns. If it is _not_ raised,
    an `AssertionError` is raised.

    :param fn: A callable that can accept the provided args and kwargs and should raise
               the specified exception.
    :param args: A list of positional arguments to apply to the function.
    :param kwargs: A dictionary of named arguments to apply to the function.
    :param exc: An exception that the function should raise.
    :returns: None
    :raises AssertionError: If the specified exception was not raised while executing the function.
    """
    try:
        fn(*args, **kwargs)
    except exc:
        return
    except Exception:
        pass
    assert False, "Did not raise %s" % exc


def setup_logging(
    log_level: str,
    msg_format: Optional[str] = None,
    formatter_class_name: Optional[str] = None,
) -> None:
    """
    Sets up the logging system.

    :param log_level: The logging level to configure the logging system for. Will be applied globally.
    :param msg_format: The format to use for the logging output. (Defaults to a sane option).
    :param formatted_class_name: Optionally provide a custom log formatter class.
    """
    msg_format = (
        msg_format
        or "%(asctime)s [%(levelname)s] %(name)s.%(funcName)s:%(lineno)s » %(message)s"
    )
    configuration: Dict[str, Any] = {
        "version": 1,
        "formatters": {"default": {"format": msg_format}},
        "handlers": {
            "console": {
                "level": log_level,
                "class": "logging.StreamHandler",
                "formatter": "default",
                "stream": "ext://sys.stdout",
            },
            "file": {
                "level": log_level,
                "class": "logging.handlers.RotatingFileHandler",
                "formatter": "default",
                "filename": "error.log",
                "maxBytes": 1024 * 1024 * 1024,
            },
        },
        "root": {
            "level": log_level,
            "handlers": ["console", "file"],
            "propagate": True,
        },
        "disable_existing_loggers": False,
    }
    if formatter_class_name:
        configuration["formatters"]["default"]["class"] = formatter_class_name
    dictConfig(configuration)
    logger.debug("Logging configuration: %s", configuration)
